newm-server/src/main/kotlin/io/newm/server/features/walletconnection/repo/WalletConnectionRepositoryImpl.kt

```diff
@@ -128,27 +131,35 @@ internal class WalletConnectionRepositoryImpl(
     override suspend fun answerChallenge(request: AnswerChallengeRequest): AnswerChallengeResponse {
         logger.debug { "answerChallenge: $request" }
 
-        val challenge =
+        val (challengeId, method, stakeAddress) =
             transaction {
                 WalletConnectionChallengeEntity.deleteAllExpired(challengeTimeToLive)
-                WalletConnectionChallengeEntity[request.challengeId]
+                WalletConnectionChallengeEntity[request.challengeId].run {
+                    Triple(id.value, method, stakeAddress).also { delete() }
+                }
             }
 
-        when (challenge.method) {
-            ChallengeMethod.SignedData -> {
-                // TODO: call newm-chain to verify signed data, if verification fails throw HttpForbiddenException
-            }
+        val response =
+            when (method) {
+                ChallengeMethod.SignedData -> {
+                    requireNotNull(request.key) { "Missing key" }
+                    cardanoRepository.verifySignData(request.signature, request.key)
+                }
 
-            ChallengeMethod.SignedTransaction -> {
-                // TODO: call newm-chain to verify signed transaction, if verification fails throw HttpForbiddenException
+                ChallengeMethod.SignedTransaction -> {
+                    cardanoRepository.verifySignTransaction(request.signature)
+                }
             }
+
+        if (!response.verified || response.challenge != buildChallengeString(challengeId, stakeAddress)) {
+            throw HttpForbiddenException("Challenge signature verification failed")
         }
 
         // we're ready to connect!!!
         val connection =
             transaction {
                 WalletConnectionEntity.new {
-                    stakeAddress = challenge.stakeAddress
+                    this.stakeAddress = stakeAddress
```

Comment:
How is this stake address linked to the user account?

Reply:
After the mobile device scan the QRCode they make another call which is JWT-guarded and at that point we add the userId to `WalletConnectionEntity`. It's at the connect() function (bellow).